controlplane/kubeadm/internal/workload_cluster_conditions.go

```diff
@@ -0,0 +1,532 @@
+limitations under the License.
+*/
+
+package internal
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/controlplane/kubeadm/internal/etcd"
+	etcdutil "sigs.k8s.io/cluster-api/controlplane/kubeadm/internal/etcd/util"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// UpdateEtcdConditions is responsible for updating machine conditions reflecting the status of all the etcd members.
+// This operation is best effort, in the sense that in case of problems in retrieving member status, it sets
+// the condition to Unknown state without returning any error.
+func (w *Workload) UpdateEtcdConditions(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, machines []*clusterv1.Machine) {
+	isEtcdManaged := kcp.Spec.KubeadmConfigSpec.ClusterConfiguration == nil || kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.Etcd.External == nil
+	if isEtcdManaged {
+		w.updateStakedEtcdConditions(ctx, kcp, machines)
+		return
+	}
+	w.updateExternalEtcdConditions(ctx, kcp, machines)
+}
+
+func (w *Workload) updateExternalEtcdConditions(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, machines []*clusterv1.Machine) { //nolint:unparam
+	// When KCP is not responsible for external etcd, we are reporting only health at KCP level.
+	conditions.MarkTrue(kcp, controlplanev1.EtcdClusterHealthyCondition)
+
+	// TODO: check external etcd for alarms an possibly also for member errors
+	// this requires implementing an new type of etcd client generator given that it is not possible to use nodes
+	// as a source for the etcd endpoint address; the address of the external etcd should be available on the kubeadm configuration.
+}
+
+func (w *Workload) updateStakedEtcdConditions(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, machines []*clusterv1.Machine) {
+	// NOTE: This methods uses control plane nodes only to get in contact with etcd but then it relies on etcd
+	// as ultimate source of truth for the list of members and for their health.
+	controlPlaneNodes, err := w.getControlPlaneNodes(ctx)
+	if err != nil {
+		conditions.MarkUnknown(kcp, controlplanev1.EtcdClusterHealthyCondition, controlplanev1.EtcdClusterInspectionFailedReason, "Failed to list nodes which are hosting the etcd members")
+		for _, m := range machines {
+			conditions.MarkUnknown(m, controlplanev1.MachineEtcdMemberHealthyCondition, controlplanev1.EtcdMemberInspectionFailedReason, "Failed to get the node which is hosting the etcd member")
+		}
+		return
+	}
+
+	// Update conditions for etcd members on the nodes.
+	var additionalKCPErrors []string
```

Comment:
Should this be just `errors`?

Reply:
I prefer to make it explicit those errors go on KCP; also, those errors are on top of the summary of the machine failures...

Reply:
shortened to kcpErrors